### PR TITLE
[DDW-1152] Implement acceptance tests for custom number, date and time formats

### PR DESCRIPTION
**Pull request by daniloprates**
This PR implements acceptance tests for custom number, date and time formats.

## Todos

 - [x] Initial profile setup is working as expected (user needs to select number/date/time format)
 - [x] Changing number/date/time format on the "Settings > General" screen works as expected
 - [x] All dates are displayed in the correct format based on user preference
 - [x] All numbers are displayed in the correct format based on user preference (including wallet sidebar wallet balance display)
 - [x] Users can create transactions regardless of their number format preference (including testing amounts bigger than 1 million ada)

## Screenshots

image

---

## Testing Checklist

- [Slack QA thread](https://input-output-rnd.slack.com/archives/GGKFXSKC6/p1583158551081000)
- [ ] Test

---

### Basics

- [ ] PR has been assigned and has appropriate labels (`feature`/`bug`/`chore`, `release-x.x.x`)
- [ ] PR is updated to the most recent version of the target branch (and there are no conflicts)
- [ ] PR has a good description that summarizes all changes and shows some screenshots or animated GIFs of important UI changes
- [ ] CHANGELOG entry has been added to the top of the appropriate section (*Features*, *Fixes*, *Chores*) and is linked to the correct PR on GitHub
- [ ] Automated tests: All acceptance and unit tests are passing (`yarn test`)
- [ ] Manual tests (minimum tests should cover newly added feature/fix): App works correctly in *development* build (`yarn dev`)
- [ ] Manual tests (minimum tests should cover newly added feature/fix): App works correctly in *production* build (`yarn package` / CI builds)
- [ ] There are no *flow* errors or warnings (`yarn flow:test`)
- [ ] There are no *lint* errors or warnings (`yarn lint`)
- [ ] There are no *prettier* errors or warnings (`yarn prettier:check`)
- [ ] There are no missing translations (running `yarn manage:translations` produces no changes)
- [ ] Text changes are proofread and approved (Jane Wild / Amy Reeve)
- [ ] Japanese text changes are proofread and approved (Junko Oda)
- [ ] UI changes look good in all themes (Alexander Rukin)
- [ ] Storybook works and no stories are broken (`yarn storybook`)
- [ ] In case of dependency changes `yarn.lock` file is updated

### Code Quality
- [ ] Important parts of the code are properly commented and documented
- [ ] Code is properly typed with flow
- [ ] React components are split-up enough to avoid unnecessary re-renderings
- [ ] Any code that only works in main process is neatly separated from components

### Testing
- [ ] New feature/change is covered by acceptance tests
- [ ] New feature/change is manually tested and approved by QA team
- [ ] All existing acceptance tests are still up-to-date
- [ ] New feature/change is covered by Daedalus Testing scenario
- [ ] All existing Daedalus Testing scenarios are still up-to-date

### After Review
- [ ] Merge the PR
- [ ] Delete the source branch
- [ ] Move the ticket to `done` column on the YouTrack board
- [ ] Update Slack QA thread by marking it with a green checkmark

### CODE REVIEW

File: tests/settings/e2e/steps/custom-formats.js [hunk 1]
```diff
@@ -0,0 +1,90 @@
...
+
+Then(/^the "([^"]*)" should display the "([^"]*)" of value "([^"]*)"$/, async function(screenElement, expectedParam, expectedValue) {
+  const currentValue = await getValueFromSelector.call(this, screenElement, expectedParam)
+  expect(currentValue).to.equal(expectedValue);
+});
+
+When(/^the "([^"]*)" wallet has received the transaction amount$/, async function(walletName) {
+  await this.client.waitUntil(async () => {
+    const currentAmount = await getValueFromSelector.call(this, walletName, 'number')
+    console.log('currentAmount', currentAmount);
```
Comment (yakovkaravelov) [inline tests/settings/e2e/steps/custom-formats.js]: @daniloprates Please remove this `console.log`

File: tests/settings/e2e/features/custom-formats.feature [hunk 2]
```diff
@@ -0,0 +1,157 @@
+@e2e  @watch
```
Comment (yakovkaravelov) [inline tests/settings/e2e/features/custom-formats.feature]: @daniloprates Please remove `@watch` tag.

**Comment by nikolaglumac** [comment]
@tomothespian please review this one 🙏

File: tests/settings/e2e/steps/custom-formats.js (hunk 1 above)
Comment (daniloprates) [reply tests/settings/e2e/steps/custom-formats.js]: @yakovkaravelov ✅

File: tests/settings/e2e/features/custom-formats.feature (hunk 2 above)
Comment (daniloprates) [reply tests/settings/e2e/features/custom-formats.feature]: @yakovkaravelov ✅

**Comment by daniloprates** [comment]
> @daniloprates Great work, just left minor comments. 👍

Thanks @yakovkaravelov. Didn't realize `lint` doesn't catch console.logs inside tests.

Fixed ✅

File: tests/settings/e2e/steps/custom-formats.js [hunk 3]
```diff
@@ -0,0 +1,89 @@
...
+
+Given(/^I have changed the following custom formats:$/, async function(formatsTable) {
+  const chosenFormats = formatsTable.hashes();
+  await this.client.executeAsync((chosenFormats, done) => {
+    Promise.all(
+      chosenFormats.map(({ param, value }) =>
+        daedalus.stores.profile._updateUserLocalSetting({ param: `${param}Format`, value }))
+    ).then(done);
+  }, chosenFormats);
+  await timeout(1500);
```
Comment (tomothespian) [inline tests/settings/e2e/steps/custom-formats.js]: @daniloprates is `timeout()` necessary here?

File: tests/settings/e2e/steps/custom-formats.js [hunk 4]
```diff
@@ -0,0 +1,89 @@
...
+
+Then(/^the "([^"]*)" should display the "([^"]*)" of value "([^"]*)"$/, async function(screenElement, expectedParam, expectedValue) {
+  const currentValue = await getValueFromSelector.call(this, screenElement, expectedParam)
+  expect(currentValue).to.equal(expectedValue);
+});
+
+When(/^the "([^"]*)" wallet has received the transaction amount$/, async function(walletName) {
+  await this.client.waitUntil(async () => {
+    const currentAmount = await getValueFromSelector.call(this, walletName, 'number')
+    return currentAmount !== '0 ADA';
```
Comment (tomothespian) [inline tests/settings/e2e/steps/custom-formats.js]: @daniloprates maybe here would be better to check that displaied `currenAmount` is equal to transaction amount from the store

File: tests/settings/e2e/steps/helpers.js [hunk 5]
```diff
@@ -89,3 +126,69 @@ export const termsOfUseHelpers = {
...
+  return {
+    selectedNumber,
+    selectedDate,
+    selectedTime,
+  }
+}
+
+export const getValueFromSelector = async function  (screenElement: string, expectedParam: string) {
+  const selector = screenElementSelectors[screenElement][expectedParam];
+  // await this.client.waitForVisible(selector);
```
Comment (tomothespian) [inline tests/settings/e2e/steps/helpers.js]: @daniloprates please remove the comment

File: source/renderer/app/components/wallet/transactions/Transaction.js [hunk 6]
```diff
@@ -379,7 +379,7 @@ export default class Transaction extends Component<Props, State> {
               <div className={styles.details}>
                 <div className={styles.type}>
                   {intl.formatMessage(messages.type, { currency })},{' '}
-                  {moment(data.date).format(currentTimeFormat)}
+                  <em>{moment(data.date).format(currentTimeFormat)}</em>
```
Comment (tomothespian) [inline source/renderer/app/components/wallet/transactions/Transaction.js]: @daniloprates maybe we shouldn't change code that we want to test just to fit tests. Is this some requirement not related to tests?
If you need that for test and if is impossible to get that in scenario rather use React.Fragment than <em> which will transform text!

**Comment by tomothespian** [comment]
@daniloprates I got this few times. Sometimes pass and sometimes not.
Screen Shot 2020-03-03 at 13 21 25

File: tests/settings/e2e/steps/custom-formats.js (hunk 4 above)
Comment (daniloprates) [reply tests/settings/e2e/steps/custom-formats.js]: ✅

File: tests/settings/e2e/steps/custom-formats.js (hunk 3 above)
Comment (daniloprates) [reply tests/settings/e2e/steps/custom-formats.js]: ✅

File: tests/settings/e2e/steps/helpers.js (hunk 5 above)
Comment (daniloprates) [reply tests/settings/e2e/steps/helpers.js]: ✅

File: source/renderer/app/components/wallet/transactions/Transaction.js (hunk 6 above)
Comment (daniloprates) [reply source/renderer/app/components/wallet/transactions/Transaction.js]: ✅

**Comment by daniloprates** [comment]
> @daniloprates Amazing work. 🚀
> I left just a few comments. Can you please check?
> Also, when I run all tests together I get one failed test but not related to this task. I will run again and let you know!

@tomothespian thanks for the review and requests. I've fixed and adjusted them.

**Comment by tomothespian** [comment]
@daniloprates Nice improvements, thanks.
But I still get this error:
Screen Shot 2020-03-03 at 15 23 29

**Comment by daniloprates** [comment]
> @daniloprates Nice improvements, thanks.
> But I still get this error:
> Screen Shot 2020-03-03 at 15 23 29

@tomothespian I could never reproduce it 😕 but made some adjustments to try avoiding it.

**Comment by tomothespian** [comment]
@daniloprates hmm weird. I will try one more time.

**Comment by tomothespian** [comment]
@daniloprates it is really weird. When I run just run your new tests with **@watch** flag all tests pass but when I run all tests together then always fails on the step I reported above :(

**Comment by daniloprates** [comment]
> @daniloprates it is really weird. When I just run your new tests with **@watch** flag all tests pass but when I run all tests together then always fails on the step I reported above :(

Fixed ✅

**Comment by tomothespian** [comment]
@daniloprates thanks. Now your tests works but when I run all tests together always fails on same scenario:
Screen Shot 2020-03-04 at 10 41 05

Can you please check?

**Comment by daniloprates** [comment]
> @daniloprates thanks. Now your tests works but when I run all tests together always fails on same scenario:
> Can you please check?

@tomothespian fixed ✅ I'll surely go into more detail on my current task `DDW-1213 Optimize E2E tests`.

**Comment by nikolaglumac** [comment]
@tomothespian can you please re-test this one? 🙏

**Comment by tomothespian** [comment]
@daniloprates same thing happened again for step: `Successfully transferring funds from "Daedalus Balance" wallet to "Daedalus Rewards" wallet`

**Comment by daniloprates** [comment]
> @daniloprates same thing happened again for step: `Successfully transferring funds from "Daedalus Balance" wallet to "Daedalus Rewards" wallet`

@tomothespian this test has failed numerous times on `v2-integration` as well, for me. I'm working on "DDW-1213 Optimize E2E tests" and will tackle this and other ones that are unrelated to `DDW-1868`.

**Comment by nikolaglumac** [comment]
@daniloprates please merge latest v2-integration - it looks like tests are more stable with the latest version of the cardano-wallet and Jormungandr which is available in v2-integration

**Comment by daniloprates** [comment]
> @daniloprates please merge latest v2-integration - it looks like tests are more stable with the latest version of the cardano-wallet and Jormungandr which is available in v2-integration

@nikolaglumac the tests I'm fixing now here are not related to this PR. I'd rather merge this one and fix them on "DDW-1213 Optimize E2E tests" if it's ok.

**Pull request merged**